opendbc/can/parser_pyx.pyx

```diff
@@ -6,12 +6,64 @@ from libcpp.string cimport string
 from libcpp.vector cimport vector
 from libc.stdint cimport uint32_t
 
-from .common cimport CANParser as cpp_CANParser
-from .common cimport dbc_lookup, Msg, DBC, CanData
+from .common cimport CANParser as cpp_CANParser, MessageState as cpp_MessageState
+from .common cimport dbc_lookup, DBC, CanData
 
 import numbers
+from collections.abc import Mapping
 from collections import defaultdict
 
+cdef class MessageState:
+  cdef cpp_MessageState *_state
+  cdef list _signal_names
+
+  def value(self, key):
+    return self._state.signal_values.at(key).value
+
+  def ts_nanos(self, key):
+    return self._state.signal_values.at(key).ts_nanos
+
+  def all_values(self, key):
+    return self._state.signal_values.at(key).all_values
+
+  @property
+  def signal_names(self):
+    return self._signal_names
+
+  @staticmethod
+  cdef create(cpp_MessageState *state):
+    message_state = MessageState()
+    message_state._state = state
+    message_state._signal_names = [it.first.decode("utf-8") for it in state.signal_values]
+    return message_state
+
+
+class ReadonlyDict(Mapping):
+  def __init__(self, state):
+    self.state = state
+    self.keys = self.state.signal_names
+
+  def __iter__(self):
+    return iter(self.keys)
+
+  def __len__(self):
+    return len(self.keys)
+
+
+class ValueDict(ReadonlyDict):
+  def __getitem__(self, key):
+    return self.state.value(key)
+
+
+class NanosDict(ReadonlyDict):
+  def __getitem__(self, key):
+    return self.state.ts_nanos(key)
+
+
+class AllValueDict(ReadonlyDict):
+  def __getitem__(self, key):
+    return self.state.all_values(key)
```

Comment:
Anyway we can get by with `ReadonlyDict`?

Reply:
how about [ebf4696](https://github.com/commaai/opendbc/pull/1439/commits/ebf4696481c6aef07319011589c8715012c825cd)